dipy/align/_public.py

```diff
@@ -0,0 +1,833 @@
+    template. If reg_method is "syn", the second is a DiffeomorphicMap class
+    instance that can be used to transform between the two spaces. Otherwise,
+    if reg_method is "aff", this is a 4x4 matrix encoding the affine transform.
+
+    Notes
+    -----
+    This function assumes that the DWI data is already internally registered.
+    See :func:`register_dwi_series`.
+
+    """
+    dwi_data, dwi_affine = read_img_arr_or_path(dwi, affine=dwi_affine)
+
+    if template is None:
+        template = dpd.read_mni_template()
+
+    template_data, template_affine = read_img_arr_or_path(
+                                       template,
+                                       affine=template_affine)
+    if isinstance(dwi, str):
+        dwi = nib.load(dwi)
+
+    if not isinstance(gtab, dpg.GradientTable):
+        gtab = dpg.gradient_table(*gtab)
+
+    mean_b0 = np.mean(dwi_data[..., gtab.b0s_mask], -1)
+    if reg_method == "syn":
+        warped_b0, mapping = syn_registration(mean_b0, template_data,
+                                              moving_affine=dwi_affine,
+                                              static_affine=template_affine,
+                                              **reg_kwargs)
+    elif reg_method == "aff":
+        warped_b0, mapping = affine_registration(mean_b0, template_data,
+                                                 moving_affine=dwi_affine,
+                                                 static_affine=template_affine,
+                                                 **reg_kwargs)
+    return warped_b0, mapping
+
+
+def write_mapping(mapping, fname):
+    """
+    Write out a syn registration mapping to a nifti file
+
+    Parameters
+    ----------
+    mapping : a DiffeomorphicMap object derived from :func:`syn_registration`
+    fname : str
+        Full path to the nifti file storing the mapping
+
+    Notes
+    -----
+    The data in the file is organized with shape (X, Y, Z, 2, 3, 3), such
+    that the forward mapping in each voxel is in `data[i, j, k, 0, :, :]` and
+    the backward mapping in each voxel is in `data[i, j, k, 0, :, :]`.
+    """
+    mapping_data = np.array([mapping.forward.T, mapping.backward.T]).T
+    nib.save(nib.Nifti1Image(mapping_data, mapping.codomain_world2grid), fname)
+
+
+def read_mapping(disp, domain_img, codomain_img, prealign=None):
+    """
+    Read a syn registration mapping from a nifti file
+
+    Parameters
+    ----------
+    disp : str or Nifti1Image
+        A file of image containing the mapping displacement field in each voxel
+        Shape (x, y, z, 3, 2)
+
+    domain_img : str or Nifti1Image
+
+    codomain_img : str or Nifti1Image
+
+    Returns
+    -------
+    A :class:`DiffeomorphicMap` object.
+
+    Notes
+    -----
+    See :func:`write_mapping` for the data format expected.
+    """
+    if isinstance(disp, str):
+        disp = nib.load(disp)
+
+    if isinstance(domain_img, str):
+        domain_img = nib.load(domain_img)
```

Comment:
same comment as above for the load/save

Reply:
The problem here (and below) is that we actually don't need the data read from file here. Just the shape and affine, so I think that for these two it would make more sense to use nibabel directly.

Reply:
Is this still an issue?